ValueError: PurePosixPath('.') has an empty name

<!--
Thanks for submitting an issue!

Quick check-list while reporting bugs:
-->

- [ ] a detailed description of the bug or problem you are having
- [ ] output of `pip list` from the virtual environment you are using
- [ ] pytest and operating system versions
- [ ] minimal example if possible


I've bracketed the exception to `pytest==7.4.2`:


with:
```console
$ cat test_debug_pytest.py
def test_my_test():
    assert True
$ cat pyproject.toml
[tool.pytest.ini_options]
addopts = "-ra --import-mode=importlib"
$ cat __init__.py

$ pytest test_debug_pytest.py
================================================================================ test session starts =================================================================================
platform linux -- Python 3.8.11, pytest-7.4.2, pluggy-0.13.1
rootdir: /mnt/c/Users/.../test
configfile: pyproject.toml
plugins: forked-1.3.0, notebook-0.6.1, mock-3.6.1, anyio-3.3.0, subtests-0.9.0, timeout-2.1.0, xdist-2.3.0, dash-2.7.0, cov-4.0.0, flaky-3.7.0, asyncio-0.21.0
asyncio: mode=strict
collected 1 item

test_debug_pytest.py E                                                                                                                                                         [100%]

======================================================================================= ERRORS =======================================================================================
___________________________________________________________________________ ERROR at setup of test_my_test ___________________________________________________________________________

self = PurePosixPath('.'), suffix = '.py'

    def with_suffix(self, suffix):
        """Return a new path with the file suffix changed.  If the path
        has no suffix, add given suffix.  If the given suffix is an empty
        string, remove the suffix from the path.
        """
        f = self._flavour
        if f.sep in suffix or f.altsep and f.altsep in suffix:
            raise ValueError("Invalid suffix %r" % (suffix,))
        if suffix and not suffix.startswith('.') or suffix == '.':
            raise ValueError("Invalid suffix %r" % (suffix))
        name = self.name
        if not name:
>           raise ValueError("%r has an empty name" % (self,))
E           ValueError: PurePosixPath('.') has an empty name

/home/me/miniconda3/envs/aq/lib/python3.8/pathlib.py:872: ValueError
============================================================================== short test summary info ===============================================================================
ERROR test_debug_pytest.py::test_my_test - ValueError: PurePosixPath('.') has an empty name
================================================================================== 1 error in 0.14s ==================================================================================
```
with no `conftest.py`.

--- a/src/_pytest/pathlib.py
+++ b/src/_pytest/pathlib.py
@@ -623,8 +623,9 @@ def module_name_from_path(path: Path, root: Path) -> str:
         # Use the parts for the relative path to the root path.
         path_parts = relative_path.parts
 
-    # Module name for packages do not contain the __init__ file.
-    if path_parts[-1] == "__init__":
+    # Module name for packages do not contain the __init__ file, unless
+    # the `__init__.py` file is at the root.
+    if len(path_parts) >= 2 and path_parts[-1] == "__init__":
         path_parts = path_parts[:-1]
 
     return ".".join(path_parts)

--- a/testing/test_pathlib.py
+++ b/testing/test_pathlib.py
@@ -28,6 +28,7 @@ from _pytest.pathlib import resolve_package_path
 from _pytest.pathlib import safe_exists
 from _pytest.pathlib import symlink_or_skip
 from _pytest.pathlib import visit
+from _pytest.pytester import Pytester
 from _pytest.tmpdir import TempPathFactory
 
 
@@ -592,6 +593,10 @@ class TestImportLibMode:
         result = module_name_from_path(tmp_path / "src/app/__init__.py", tmp_path)
         assert result == "src.app"
 
+        # Unless __init__.py file is at the root, in which case we cannot have an empty module name.
+        result = module_name_from_path(tmp_path / "__init__.py", tmp_path)
+        assert result == "__init__"
+
     def test_insert_missing_modules(
         self, monkeypatch: MonkeyPatch, tmp_path: Path
     ) -> None:
@@ -662,6 +667,22 @@ class TestImportLibMode:
 
         mod = import_path(init, root=tmp_path, mode=ImportMode.importlib)
         assert len(mod.instance.INSTANCES) == 1
+
+    def test_importlib_root_is_package(self, pytester: Pytester) -> None:
+        """
+        Regression for importing a `__init__`.py file that is at the root
+        (#11417).
+        """
+        pytester.makepyfile(__init__="")
+        pytester.makepyfile(
+            """
+            def test_my_test():
+                assert True
+            """
+        )
+
+        result = pytester.runpytest("--import-mode=importlib")
+        result.stdout.fnmatch_lines("* 1 passed *")
 
 
 def test_safe_exists(tmp_path: Path) -> None:

